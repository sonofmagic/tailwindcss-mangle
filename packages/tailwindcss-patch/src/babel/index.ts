export { default as generate } from '@babel/generator'
export { default as traverse } from '@babel/traverse'
export { parse } from '@babel/parser'
