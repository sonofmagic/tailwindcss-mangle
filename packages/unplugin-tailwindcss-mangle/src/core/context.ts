import fs from 'node:fs'
import { resolve } from 'node:path'
import { ClassGenerator } from '@tailwindcss-mangle/shared'
import { getConfig, getDefaultMangleUserConfig } from '@tailwindcss-mangle/config'
import type { MangleUserConfig } from '@tailwindcss-mangle/config'
import { sort } from 'fast-sort'
import defu from 'defu'
import { createGlobMatcher, defaultMangleClassFilter } from '@/utils'

export class Context {
  options: MangleUserConfig
  includeMatcher: (file: string) => boolean
  excludeMatcher: (file: string) => boolean
  classSet: Set<string>
  replaceMap: Map<string, string>
  classGenerator: ClassGenerator
  constructor(opts: MangleUserConfig) {
    this.options = defu(opts, getDefaultMangleUserConfig())
    this.classSet = new Set()
    this.replaceMap = new Map()
    this.includeMatcher = createGlobMatcher(this.options.include, true)
    this.excludeMatcher = createGlobMatcher(this.options.exclude, false)
    this.classGenerator = new ClassGenerator(this.options.classGenerator)
  }

  mergeOptions(opts?: MangleUserConfig) {
    // 插件选项优先
    this.options = defu(this.options, opts)
    this.includeMatcher = createGlobMatcher(this.options.include, true)
    this.excludeMatcher = createGlobMatcher(this.options.exclude, false)
    this.classGenerator = new ClassGenerator(this.options.classGenerator)
  }

  isInclude(file: string) {
    return this.includeMatcher(file) && !this.excludeMatcher(file)
  }

  currentMangleClassFilter(className: string) {
    return (this.options.mangleClassFilter ?? defaultMangleClassFilter)(className)
  }

  getClassSet() {
    return this.classSet
  }

  getReplaceMap() {
    return this.replaceMap
  }

  addToUsedBy(key: string, file: string) {
    const hit = this.classGenerator.newClassMap[key]
    if (hit) {
      hit.usedBy.add(file)
    }
  }

  async initConfig() {
    const { config } = await getConfig()
    const mangleConfig = config?.mangle
    this.mergeOptions(mangleConfig)
    const jsonPath = this.options.classListPath ?? resolve(process.cwd(), config?.patch?.output?.filename as string)

    if (jsonPath && fs.existsSync(jsonPath)) {
      const rawClassList = fs.readFileSync(jsonPath, 'utf8')
      const list = JSON.parse(rawClassList) as string[]
      // why?
      // case bg-red-500 and bg-red-500/50
      // transform bg-red-500/50 first
      const classList = sort(list).desc((c) => c.length)
      for (const className of classList) {
        if (this.currentMangleClassFilter(className)) {
          this.classSet.add(className)
        }
      }
    }
    for (const cls of this.classSet) {
      this.classGenerator.generateClassName(cls)
    }

    for (const x of Object.entries(this.classGenerator.newClassMap)) {
      this.replaceMap.set(x[0], x[1].name)
    }
    return config
  }
}
