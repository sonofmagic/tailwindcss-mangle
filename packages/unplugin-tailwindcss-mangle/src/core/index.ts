export * from './plugin'
export * from './context'
