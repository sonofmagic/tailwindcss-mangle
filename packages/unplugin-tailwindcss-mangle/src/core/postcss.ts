import postcss, { type PluginCreator } from 'postcss'
import parser from 'postcss-selector-parser'

const transformSelectorPostcssPlugin: PluginCreator<{
  replaceMap: Map<string, string>
}> = function (options) {
  const { replaceMap } = options ?? {}
  return {
    postcssPlugin: 'postcssTransformSelectorPlugin',
    async Rule(rule) {
      await parser((selectors) => {
        selectors.walkClasses((s) => {
          if (s.value && replaceMap && replaceMap.has(s.value)) {
            const v = replaceMap.get(s.value)
            if (v) {
              s.value = v
            }
          }
        })
      }).transform(rule, {
        lossless: false,
        updateSelector: true
      })
    }
  }
}
transformSelectorPostcssPlugin.postcss = true

export function processCss(options: { replaceMap: Map<string, string>; css: string; file: string }) {
  const { css, file, replaceMap } = options
  return postcss([
    transformSelectorPostcssPlugin({
      replaceMap
    })
  ]).process(css, {
    from: file,
    to: file
  })
}
