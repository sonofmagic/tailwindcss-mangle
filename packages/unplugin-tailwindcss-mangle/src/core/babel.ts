import babel from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'
import MagicString from 'magic-string'
import { splitCode } from '@tailwindcss-mangle/shared'
interface Options {
  replaceMap: Map<string, string>
  magicString: MagicString
}

export const plugin = declare((api, options: Options) => {
  api.assertVersion(7)
  const { magicString, replaceMap } = options
  return {
    visitor: {
      StringLiteral: {
        enter(p) {
          const node = p.node
          let value = node.value
          const arr = splitCode(value)

          for (const str of arr) {
            if (replaceMap.has(str)) {
              const v = replaceMap.get(str)
              if (v) {
                value = value.replaceAll(str, v)
              }
            }
          }
          if (typeof node.start === 'number' && typeof node.end === 'number' && value) {
            magicString.update(node.start + 1, node.end - 1, value)
          }
        }
      }
    }
  }
})

export function processJs(options: { code: string; replaceMap: Map<string, string> }) {
  const { code, replaceMap } = options
  const magicString = new MagicString(code)
  babel.transformSync(code, {
    presets: [
      // ['@babel/preset-react', {}],
      [
        '@babel/preset-typescript',
        {
          allExtensions: true,
          isTSX: true
        }
      ]
    ],
    plugins: [
      [
        plugin,
        {
          magicString,
          replaceMap
        }
      ]
    ]
  })
  return magicString.toString()
}
