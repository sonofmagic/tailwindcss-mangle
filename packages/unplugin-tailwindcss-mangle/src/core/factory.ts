import type { TransformerOptions } from '@tailwindcss-mangle/config'
import type { UnpluginFactory } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import { Context, cssHandler, htmlHandler, jsHandler } from '@tailwindcss-mangle/core'
import { isCSSRequest } from 'is-css-request'
import path from 'pathe'
import { getGroupedEntries } from '@/utils'
import { pluginName } from '../constants'

const WEBPACK_LOADER = path.resolve(__dirname, __DEV__ ? '../../dist/loader.cjs' : './loader.cjs')

const factory: UnpluginFactory<TransformerOptions | undefined> = (options) => {
  const ctx = new Context()
  let filter = (_id: string) => true
  return [
    {
      name: `${pluginName}:pre`,
      enforce: 'pre',
      async buildStart() {
        await ctx.initConfig({
          transformerOptions: options,
        })
        filter = createFilter(ctx.options.sources?.include, ctx.options.sources?.exclude)
      },
    },
    {
      name: `${pluginName}`,
      transformInclude(id) {
        return filter(id)
      },
      async transform(code, id) {
        const opts = {
          ctx,
          id,
        }
        if (/\.[cm]?[jt]sx?(?:$|\?)/.test(id)) {
          return jsHandler(code, opts)
        }
        else if (/\.(?:vue|svelte)(?:$|\?)/.test(id)) {
          if (isCSSRequest(id)) {
            return await cssHandler(code, opts)
          }
          else {
            return jsHandler(code, opts)
          }
        }
        else if (isCSSRequest(id)) {
          return await cssHandler(code, opts)
        }
        else if (/\.html?/.test(id)) {
          return htmlHandler(code, opts)
        }
      },
      webpack(compiler) {
        const { NormalModule } = compiler.webpack
        const isExisted = true
        compiler.hooks.compilation.tap(pluginName, (compilation) => {
          NormalModule.getCompilationHooks(compilation).loader.tap(pluginName, (_loaderContext, module) => {
            if (isExisted) {
              const idx = module.loaders.findIndex(x => x.loader.includes('postcss-loader'))

              if (idx > -1) {
                module.loaders.splice(idx, 0, {
                  loader: WEBPACK_LOADER,
                  ident: null,
                  options: {
                    ctx,
                  },
                  type: null,
                })
              }
            }
          })
        })
      },
    },
    {
      name: `${pluginName}:post`,
      enforce: 'post',
      vite: {
        transformIndexHtml(html) {
          const { code } = htmlHandler(html, { ctx })
          return code
        },
        // generateBundle: {
        //   async handler(options, bundle) {
        //     const groupedEntries = getGroupedEntries(Object.entries(bundle))

        //     if (Array.isArray(groupedEntries.css) && groupedEntries.css.length > 0) {
        //       for (let i = 0; i < groupedEntries.css.length; i++) {
        //         const [id, cssSource] = groupedEntries.css[i] as [string, OutputAsset]

        //         const { code } = await cssHandler(cssSource.source.toString(), {
        //           id,
        //           ctx,
        //         })
        //         cssSource.source = code
        //       }
        //     }
        //   },
        // },
      },
      webpack(compiler) {
        const { Compilation, sources } = compiler.webpack
        const { ConcatSource } = sources

        compiler.hooks.compilation.tap(pluginName, (compilation) => {
          compilation.hooks.processAssets.tapPromise(
            {
              name: pluginName,
              stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE,
            },
            async (assets) => {
              const groupedEntries = getGroupedEntries(Object.entries(assets))

              // if (groupedEntries.js.length > 0) {
              //   for (let i = 0; i < groupedEntries.js.length; i++) {
              //     const [file, chunk] = groupedEntries.js[i]

              //     const code = jsHandler(chunk.source().toString(), {
              //       ctx,
              //     }).code
              //     if (code) {
              //       const source = new ConcatSource(code)
              //       compilation.updateAsset(file, source)
              //     }
              //   }
              // }

              if (groupedEntries.css.length > 0) {
                for (let i = 0; i < groupedEntries.css.length; i++) {
                  const [id, cssSource] = groupedEntries.css[i]

                  const { code } = await cssHandler(cssSource.source().toString(), {
                    id,
                    ctx,
                  })

                  const source = new ConcatSource(code)

                  compilation.updateAsset(id, source)
                }
              }

              // if (groupedEntries.html.length > 0) {
              //   for (let i = 0; i < groupedEntries.html.length; i++) {
              //     const [file, asset] = groupedEntries.html[i]

              //     const { code } = htmlHandler(asset.source().toString(), {
              //       ctx,
              //     })
              //     const source = new ConcatSource(code)
              //     compilation.updateAsset(file, source)
              //   }
              // }
            },
          )
        })
      },
      writeBundle() {
        ctx.dump()
      },
    },
  ]
}

export default factory
