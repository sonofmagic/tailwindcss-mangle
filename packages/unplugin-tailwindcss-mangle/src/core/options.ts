import fs from 'node:fs'
import { resolve } from 'node:path'
import { ClassGenerator } from '@tailwindcss-mangle/shared'
import { getConfig, getDefaultUserConfig } from '@tailwindcss-mangle/config'
import type { UserConfig } from '@tailwindcss-mangle/config'
import defu from 'defu'
import { sort } from 'fast-sort'
import { htmlHandler, cssHandler, jsHandler, preProcessJs } from '@tailwindcss-mangle/core'
import type { Options, ClassMapOutputOptions } from '@/types'
import { createGlobMatcher, defaultMangleClassFilter } from '@/utils'

export function getOptions(opts: Options | undefined = {}) {
  const {
    include,
    exclude,
    disabled,
    mangleClassFilter,
    classMapOutput,
    classGenerator: classGeneratorOptions,
    classListPath: _classListPath
  } = defu<Options, Options[]>(opts, {
    include: ['**/*.{js,jsx,ts,tsx,svelte,vue}'],
    exclude: ['**/*.{css,scss,less,sass,postcss,html,htm}'],
    disabled: process.env.NODE_ENV === 'development'
  })
  const includeMatcher = createGlobMatcher(include, true)
  const excludeMatcher = createGlobMatcher(exclude, false)
  const currentMangleClassFilter = mangleClassFilter ?? defaultMangleClassFilter

  function isInclude(file: string) {
    return includeMatcher(file) && !excludeMatcher(file)
  }

  const classSet: Set<string> = new Set()
  const replaceMap: Map<string, string> = new Map()
  let userConfig: UserConfig = getDefaultUserConfig()

  const classMapOutputOptions: ClassMapOutputOptions = {
    filename: 'classMap.json'
  }

  if (typeof classMapOutput === 'object') {
    Object.assign(classMapOutputOptions, classMapOutput)
  }

  // let cached: boolean
  const classGenerator = new ClassGenerator(classGeneratorOptions)

  function getCachedClassSet() {
    return classSet
  }

  function getReplaceMap() {
    return replaceMap
  }

  async function initConfig() {
    const { config } = await getConfig()
    userConfig = config as UserConfig
    let classListPath: string = ''
    if (userConfig) {
      classListPath = resolve(process.cwd(), userConfig.patch?.output?.filename as string)
    }
    if (_classListPath) {
      classListPath = _classListPath
    }

    if (classListPath && fs.existsSync(classListPath)) {
      const rawClassList = fs.readFileSync(classListPath, 'utf8')
      const list = JSON.parse(rawClassList) as string[]
      // why?
      // case bg-red-500 and bg-red-500/50
      // transform bg-red-500/50 first
      const classList = sort(list).desc((c) => c.length)
      for (const className of classList) {
        if (currentMangleClassFilter(className)) {
          classSet.add(className)
        }
      }
    }
    for (const cls of classSet) {
      classGenerator.generateClassName(cls)
    }

    for (const x of Object.entries(classGenerator.newClassMap)) {
      replaceMap.set(x[0], x[1].name)
    }
    return config
  }

  function addToUsedBy(key: string, file: string) {
    const hit = classGenerator.newClassMap[key]
    if (hit) {
      hit.usedBy.add(file)
    }
  }

  return {
    getCachedClassSet,
    classGenerator,
    includeMatcher,
    excludeMatcher,
    isInclude,
    classMapOutputOptions,
    initConfig,
    getReplaceMap,
    addToUsedBy,
    disabled,
    htmlHandler,
    cssHandler,
    jsHandler,
    preProcessJs
  }
}
