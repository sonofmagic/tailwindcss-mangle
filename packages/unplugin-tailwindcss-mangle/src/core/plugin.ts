import { dirname } from 'node:path'
import fs from 'node:fs/promises'
import { createUnplugin } from 'unplugin'
import type { OutputAsset } from 'rollup'
import { Context, cssHandler, htmlHandler, jsHandler, preProcessJs, preProcessRawCode } from '@tailwindcss-mangle/core'
import type { ClassMapOutputOptions, MangleUserConfig } from '@tailwindcss-mangle/config'
import MagicString from 'magic-string'
import { pluginName } from '@/constants'
import { ensureDir, getGroupedEntries } from '@/utils'

export const unplugin = createUnplugin((options?: MangleUserConfig) => {
  const ctx = new Context()

  return {
    name: pluginName,
    enforce: 'pre',

    async buildStart() {
      await ctx.initConfig({
        mangleOptions: options,
      })
    },
    transformInclude(id) {
      return ctx.isInclude(id)
    },
    transform(code, id) {
      const s = new MagicString(code)
      const replaceMap = ctx.getReplaceMap()
      // 直接忽略 css  文件，因为此时 tailwindcss 还没有展开
      return /\.[jt]sx?$/.test(id)
        ? preProcessJs({
          code: s,
          replaceMap,
          ctx,
          id,
        })
        : preProcessRawCode({
          code,
          ctx,
          replaceMap,
          id,
        })
    },
    vite: {
      generateBundle: {
        async handler(options, bundle) {
          const replaceMap = ctx.getReplaceMap()
          const groupedEntries = getGroupedEntries(Object.entries(bundle))

          if (Array.isArray(groupedEntries.css) && groupedEntries.css.length > 0) {
            for (let i = 0; i < groupedEntries.css.length; i++) {
              const [file, cssSource] = groupedEntries.css[i] as [string, OutputAsset]

              const { css } = await cssHandler(cssSource.source.toString(), {
                file,
                replaceMap,
                ctx,
              })
              cssSource.source = css
            }
          }
        },
      },
    },
    webpack(compiler) {
      const { Compilation, sources } = compiler.webpack
      const { ConcatSource } = sources

      compiler.hooks.compilation.tap(pluginName, (compilation) => {
        compilation.hooks.processAssets.tapPromise(
          {
            name: pluginName,
            stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE,
          },
          async (assets) => {
            const replaceMap = ctx.getReplaceMap()
            const groupedEntries = getGroupedEntries(Object.entries(assets))

            if (groupedEntries.js.length > 0) {
              for (let i = 0; i < groupedEntries.js.length; i++) {
                const [file, chunk] = groupedEntries.js[i]

                const code = jsHandler(chunk.source().toString(), {
                  replaceMap,
                  ctx,
                }).code
                if (code) {
                  const source = new ConcatSource(code)
                  compilation.updateAsset(file, source)
                }
              }
            }

            if (groupedEntries.css.length > 0) {
              for (let i = 0; i < groupedEntries.css.length; i++) {
                const [file, cssSource] = groupedEntries.css[i]

                const { css } = await cssHandler(cssSource.source().toString(), {
                  replaceMap,
                  file,
                  ctx,
                })

                const source = new ConcatSource(css)

                compilation.updateAsset(file, source)
              }
            }

            if (groupedEntries.html.length > 0) {
              for (let i = 0; i < groupedEntries.html.length; i++) {
                const [file, asset] = groupedEntries.html[i]

                const html = htmlHandler(asset.source().toString(), {
                  ctx,
                  replaceMap,
                })
                const source = new ConcatSource(html)
                compilation.updateAsset(file, source)
              }
            }
          },
        )
      })
    },
    async writeBundle() {
      if (ctx.options.classMapOutput?.enable) {
        const opts = ctx.options.classMapOutput as Required<ClassMapOutputOptions>
        const entries = Object.entries(ctx.classGenerator.newClassMap)
        if (entries.length > 0 && opts) {
          await ensureDir(dirname(opts.filename))
          const output = JSON.stringify(
            entries.map((x) => {
              return {
                origin: x[0],
                replacement: x[1].name,
                usedBy: [...x[1].usedBy],
              }
            }),
            null,
            opts.loose ? 2 : 0,
          )
          await fs.writeFile(opts.filename, output, 'utf8')
          console.log(`✨ ${opts.filename} generated!`)
        }
      }
    },
  }
})
