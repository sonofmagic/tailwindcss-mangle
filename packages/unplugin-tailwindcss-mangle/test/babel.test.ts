import path from 'node:path'
import fs from 'node:fs/promises'
import { processJs } from '@/core/babel'
import { getOptions } from '@/core/options'
const fixturesRoot = path.resolve(__dirname, './fixtures')
const tsxRoot = path.resolve(fixturesRoot, './tsx')

describe('babel plugin', () => {
  it('tsx app0', async () => {
    const { initConfig, getReplaceMap, addToUsedBy } = getOptions({
      classListPath: path.resolve(tsxRoot, './.tw-patch/tw-class-list.json')
    })
    await initConfig()
    const replaceMap = getReplaceMap()
    const file = path.resolve(tsxRoot, 'app0.tsx')
    const code = await fs.readFile(file, 'utf8')
    const res = processJs({
      code,
      replaceMap,
      addToUsedBy,
      id: file
    })
    expect(res).toMatchSnapshot()
  })
})
