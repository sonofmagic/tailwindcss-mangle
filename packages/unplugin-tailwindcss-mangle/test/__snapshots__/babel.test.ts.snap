// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel plugin > tsx app0 1`] = `
"import { useState } from 'react'
import reactLogo from './assets/react.svg'

import './App.css'

function App() {
  const [count, setCount] = useState(0)

  return (
    <main className=\\"flex tw-fb tw-zb tw-cb tw-ra tw-nc\\">
      <div className=\\"tw-sc tw-hc tw-vb tw-cb tw-ra tw-pb tw-dc tw-bc\\">
        <p className=\\"fixed tw-ec tw-kc flex tw-hc tw-va tw-xb tw-pa tw-ja tw-za tw-oc tw-pc tw-ca tw-l tw-v tw-ea tw-tb tw-ub  tw-ab tw-rb tw-wa tw-gc lg:tw-v\\">
          Get started by editing&nbsp;
          <code className=\\"tw-pb tw-ob\\">pages/index.tsx</code>
        </p>
        <div className=\\"fixed tw-yb tw-ec flex tw-lc tw-hc tw-qb tw-va tw-ka tw-jb tw-wb tw-qa tw-ua tw-tb tw-sb tw-ub tw-kb\\">
          <a
            className=\\"tw-w flex tw-ba tw-ic tw-uc tw-n tw-fc\\"
            href=\\"https://vercel.com?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            By <img src=\\"/vite.svg\\" alt=\\"Vercel Logo\\" className=\\"tw-hb\\" width={100} height={24} />
          </a>
        </div>
      </div>

      <div className=\\"relative flex tw-ba tw-na tw-ha tw-ia tw-k tw-u before:bg-gradient-radial tw-da tw-p tw-oa tw-s tw-sa tw-gb tw-la tw-ma tw-o after:bg-gradient-conic tw-y tw-z tw-ta tw-x tw-b tw-d tw-h tw-j tw-g tw-t\\">
        <img className=\\"relative tw-a tw-hb\\" src={reactLogo} alt=\\"Next.js Logo\\" width={180} height={37} />
      </div>

      <div className=\\"tw-jc grid tw-ib tw-cc tw-xa tw-db\\">
        <a
          href=\\"https://nextjs.org/docs?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app\\"
          className=\\"group tw-mb border tw-aa tw-qc tw-rc tw-ga tw-q tw-fa tw-c tw-e\\"
          target=\\"_blank\\"
          rel=\\"noopener noreferrer\\"
        >
          <h2 className={\`  tw-mc tw-ac tw-yad\`}>
            Docs <span className=\\"tw-bb tw-r tw-i tw-f\\">-&gt;</span>
          </h2>
          <p className={\`  tw-tc tw-eb tw-dc tw-lb0\`}>Find in-depth information about Next.js features and API.</p>
        </a>

        <a
          href=\\"https://nextjs.org/learn?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app\\"
          className=\\"group tw-mb border tw-aa tw-qc tw-rc tw-ga tw-q tw-fa tw-c tw-e\\"
          target=\\"_blank\\"
          rel=\\"noopener noreferrer\\"
        >
          <h2 className={\`  tw-mc tw-ac tw-yad\`}>
            Learn <span className=\\"tw-bb tw-r tw-i tw-f\\">-&gt;</span>
          </h2>
          <p className={\`  tw-tc tw-eb tw-dc tw-lb0\`}>Learn about Next.js in an interactive course with&nbsp;quizzes!</p>
        </a>

        <a
          href=\\"https://vercel.com/templates?framework=next.js&utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app\\"
          className=\\"group tw-mb border tw-aa tw-qc tw-rc tw-ga tw-q tw-fa tw-c tw-e\\"
          target=\\"_blank\\"
          rel=\\"noopener noreferrer\\"
        >
          <h2 className={\`  tw-mc tw-ac tw-yad\`}>
            Templates <span className=\\"tw-bb tw-r tw-i tw-f\\">-&gt;</span>
          </h2>
          <p className={\`  tw-tc tw-eb tw-dc tw-lb0\`}>Discover and deploy boilerplate example Next.js&nbsp;projects.</p>
        </a>

        <a
          href=\\"https://vercel.com/new?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app\\"
          className=\\"group tw-mb border tw-aa tw-qc tw-rc tw-ga tw-q tw-fa tw-c tw-e\\"
          target=\\"_blank\\"
          rel=\\"noopener noreferrer\\"
        >
          <h2 className={\`  tw-mc tw-ac tw-yad\`}>
            Deploy <span className=\\"tw-bb tw-r tw-i tw-f\\">-&gt;</span>
          </h2>
          <p className={\`  tw-tc tw-eb tw-dc tw-lb0\`}>Instantly deploy your Next.js site to a shareable URL with Vercel.</p>
        </a>
      </div>
    </main>
  )
}

export default App"
`;
